pkg/daemon/task.go

```diff
@@ -0,0 +1,61 @@
+package daemon
+
+import "time"
+
+const (
+	TaskStateRequested TaskState = iota // when client requests a task
```

Comment:
I think as a start having only `Requested`, `In progress` and `Completed` would be sufficient for both `build` and `run` jobs. At the moment Testground doesn't have `Requested`, all jobs go directly to `In progress`, and the way we detect `Completed` is by counting the containers on a given environment.

Reply:
We could definitely get rid of the `requested` state. What I was thinking is that clients submit a task in the `requested` state to the HTTP server, which pushes the task onto the queue and then responds with the task `scheduled` state.

Later, the client can get an update about the state of the task, which might be in any of the states (other than requested) -- For the server to render an update to the client, it's a simple matter of reading it from the database.

We could remove separate building and running states too;  that probably is more detail than necessary.

I wonder if we need a `failed-to-run` state of some kind to indicate build failures, etc.

Reply:
I suggest we start simple and don't have a `failed-to-run` - that would be `in progress` going to `completed` after the global timeout (for example if you try to schedule too many testplan instances, and the cluster doesn't have autoscaler setup).

1. requested / scheduled (added to the queue, but not processed) - i don't think we need a separate states here

2. in progress / processing (popped from the queue, and building or running or both, depending on the request)

3. completed (no longer running (could be successful or failed or any other terminal state) ; outputs are ready to be processed; note that at the moment we don't have history on failed tasks (for example is a pod goes OOM and we don't have keep_alive, we won't find out - we need to think about that).

Reply:
I agree with the state machine being modelled here. `Completed` should represent the terminal state, and we should probably model a `result` field with various outcomes:

* success
* failed

For the failed state, we probably want to have an `error` field in the data model, so that we can save the error for display.

Reply:
@raulk @nonsesne updated the model. Let me know what you think!